fish-rust/src/builtins/string/split.rs

```diff
@@ -0,0 +1,285 @@
+    fn from(_: crate::wutil::wcstoi::Error) -> Self {
+        FieldParseError::Number
+    }
+}
+
+impl<'args> TryFrom<&'args wstr> for Fields {
+    type Error = FieldParseError;
+
+    /// FIELDS is a comma-separated string of field numbers and/or spans.
+    /// Each field is one-indexed.
+    fn try_from(value: &wstr) -> Result<Self, Self::Error> {
+        fn parse_field(f: &wstr) -> Result<Vec<usize>, FieldParseError> {
+            use FieldParseError::*;
+            let range: Vec<&wstr> = f.split('-').collect();
+            let range: Vec<usize> = match range[..] {
+                [s, e] => {
+                    let start = fish_wcstoi(s)? as usize;
+                    let end = fish_wcstoi(e)? as usize;
+
+                    if start == 0 || end == 0 {
+                        return Err(Range);
+                    }
+
+                    if start <= end {
+                        // we store as 0-indexed, but the range is 1-indexed
+                        (start - 1..end).collect()
+                    } else {
+                        // this is allowed
+                        (end - 1..start).rev().collect()
+                    }
+                }
+                _ => match fish_wcstoi(f)? as usize {
+                    n @ 1.. => vec![n - 1],
+                    _ => return Err(Field),
+                },
+            };
+            Ok(range)
+        }
+
+        let fields = value.split(',').map(parse_field);
+
+        let mut indices = Vec::new();
+        for field in fields {
+            indices.extend(field?);
+        }
+
+        Ok(Self(indices))
+    }
+}
+
+impl<'args> StringSubCommand<'args> for Split<'args> {
+    const LONG_OPTIONS: &'static [woption<'static>] = &[
+        wopt(L!("quiet"), no_argument, 'q'),
+        wopt(L!("right"), no_argument, 'r'),
+        wopt(L!("max"), required_argument, 'm'),
+        wopt(L!("no-empty"), no_argument, 'n'),
+        wopt(L!("fields"), required_argument, 'f'),
+        // FIXME: allow-empty is not documented
+        wopt(L!("allow-empty"), no_argument, 'a'),
+    ];
+    const SHORT_OPTIONS: &'static wstr = L!(":qrm:nf:a");
+
+    fn parse_opt(&mut self, name: &wstr, c: char, arg: Option<&wstr>) -> Result<(), StringError> {
+        match c {
+            'q' => self.quiet = true,
+            'r' => self.split_from = Direction::Right,
+            'm' => {
+                self.max = fish_wcstol(arg.unwrap())?
+                    .try_into()
+                    .map_err(|_| invalid_args!("%ls: Invalid max value '%ls'\n", name, arg))?
+            }
+            'n' => self.no_empty = true,
+            'f' => {
+                self.fields = arg.unwrap().try_into().map_err(|e| match e {
+                    FieldParseError::Number => StringError::NotANumber,
+                    FieldParseError::Range => {
+                        invalid_args!("%ls: Invalid range value for field '%ls'\n", name, arg)
```

Comment:
These should now all support translation — how we extract the translation strings I leave as a future work as I don't know whether that will happen before/after macro-expansion